### PR TITLE
Api 3423 refactor token handler tests

### PR DESCRIPTION
**Pull request by lukemarkwordtlibertyits**
Created this story after Token handler was broken into a strategy pattern. Thought it would be better to not test injected dependencies for this module.
## AC
All AC met (to an extent)
- [x] A separate test module should be created for each TokenHandler module, client or strategy.
- [x] All injected dependencies should be 
- [x] 100% (wow!) code coverage for the TokenHandler Pattern.
    - Didn't quite think this one through. Line and statement coverage is at %100 percent for all modules. Case coverage is mostly at %100 but a few modules are in the 90s. 
- [x] All reused test logic should be broken into a utility module.

## Concerns

I really had my OOP hat on when I created this story. I am starting to think that while unit testing each module instead of higher up modules (with non mocked dependencies) will lead to higher code coverage, simpler tests (for the most part) and more corner case testing, it may also lead to some false positives, or silent failure of the code base. There is also the issue of refactorability. [Here](https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a) is the blog post that made me think of this. 

Before this PR unit testing strategies was a bit hit or miss. Now it is completely consistent. The integration tests, will also mitigate some of the risk of too much mocking. I'm not saying we shouldn't refactor these tests, but I am open to a discussion and a rework of this PR if necessary. Let me know your thoughts.

## Coverage Before Changes

File                                                                            | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                         
--------------------------------------------------------------------------------|---------|----------|---------|---------|-----------------------------------------------------------                                                                                         
  tokenHandler.js                                                               |     100 |      100 |     100 |     100 |                                                                
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses                          |   96.05 |    91.89 |     100 |   96.05 |                                                           
  tokenHandlerClient.js                                                         |   90.32 |    84.21 |     100 |   90.32 | 48-49,79                                                  
  tokenHandlerClientBuilder.js                                                  |     100 |      100 |     100 |     100 |                                                           
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies |   94.12 |       50 |     100 |   94.12 |                                                           
  getPatientInfoFromLaunchStrategy.js                                           |     100 |      100 |     100 |     100 |                                                           
  getPatientInfoFromValidateEndpointStrategy.js                                 |   92.86 |       50 |     100 |   92.86 | 18                                                        
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/pullDocumentStrategies   |     100 |      100 |     100 |     100 |                                                           
  pullDocumentByAccessTokenStrategy.js                                          |     100 |      100 |     100 |     100 |                                                           
  pullDocumentByCodeStrategy.js                                                 |     100 |      100 |     100 |     100 |                                                           
  pullDocumentByLaunchStrategy.js                                               |     100 |      100 |     100 |     100 |                                                           
  pullDocumentByRefreshTokenStrategy.js                                         |     100 |      100 |     100 |     100 |                                                           
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies   |     100 |    83.33 |     100 |     100 |                                                           
  saveDocumentLaunchStrategy.js                                                 |     100 |      100 |     100 |     100 |                                                           
  saveDocumentStateStrategy.js                                                  |     100 |       75 |     100 |     100 | 13                                                        
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies          |   92.59 |    72.73 |     100 |   92.59 |                                                           
  authorizationCodeStrategy.js                                                  |     100 |      100 |     100 |     100 |                                                           
  clientCredentialsStrategy.js                                                  |   86.67 |       70 |     100 |   86.67 | 30-34,48,69                                               
  refreshTokenStrategy.js                                                       |   94.59 |       75 |     100 |   94.59 | 42,59                                                     
  unsupportedGrantStrategy.js                                                   |     100 |      100 |     100 |     100 |                                                                                                   

## Coverage After Changes

File                                                                            | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                         
--------------------------------------------------------------------------------|---------|----------|---------|---------|-----------------------------------------------------------
  tokenHandler.js                                                               |     100 |      100 |     100 |     100 |                                                              
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses                          |     100 |     97.3 |     100 |     100 |                                                           
  tokenHandlerClient.js                                                         |     100 |    94.74 |     100 |     100 | 63                                                        
  tokenHandlerClientBuilder.js                                                  |     100 |      100 |     100 |     100 |                                                           
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/documentStrategies       |     100 |      100 |     100 |     100 |                                                           
  getDocumentByAccessTokenStrategy.js                                           |     100 |      100 |     100 |     100 |                                                           
  getDocumentByCodeStrategy.js                                                  |     100 |      100 |     100 |     100 |                                                           
  getDocumentByLaunchStrategy.js                                                |     100 |      100 |     100 |     100 |                                                           
  getDocumentByRefreshTokenStrategy.js                                          |     100 |      100 |     100 |     100 |                                                           
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/getPatientInfoStrategies |     100 |      100 |     100 |     100 |                                                           
  getPatientInfoFromLaunchStrategy.js                                           |     100 |      100 |     100 |     100 |                                                           
  getPatientInfoFromValidateEndpointStrategy.js                                 |     100 |      100 |     100 |     100 |                                                           
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/saveDocumentStrategies   |     100 |      100 |     100 |     100 |                                                           
  saveDocumentLaunchStrategy.js                                                 |     100 |      100 |     100 |     100 |                                                           
  saveDocumentStateStrategy.js                                                  |     100 |      100 |     100 |     100 |                                                           
 oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies          |     100 |    96.15 |     100 |     100 |                                                           
  authorizationCodeStrategy.js                                                  |     100 |      100 |     100 |     100 |                                                           
  clientCredentialsStrategy.js                                                  |     100 |      100 |     100 |     100 |                                                           
  refreshTokenStrategy.js                                                       |     100 |    91.67 |     100 |     100 | 68                                                        
  unsupportedGrantStrategy.js                                                   |     100 |      100 |     100 |     100 |

### CODE REVIEW

**Review by bdgrider — CHANGES_REQUESTED** [review]
Checked branch, all unit and regression tests are passing.

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ class ClientCredentialsStrategy {
       }
     } catch (error) {
       rethrowIfRuntimeError(error);
-      if (error.response.status == 400) {
+      if (error.response != undefined && error.response.status == 400) {
```
Comment (crolarlibertyva) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js:40]: I would switch to using something like `if (!error && error.response.status == 400)` to foolproof against a future change that would have `error.response ==- null` or `error.response === false`.

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js [hunk 2]
```diff
@@ -51,7 +51,7 @@ class ClientCredentialsStrategy {
             error_description: error.response.data.error_description,
           };
         }
-      } else if (error.response.status == 401) {
+      } else if (error.response != undefined && error.response.status == 401) {
```
Comment (crolarlibertyva) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js:54]: Same pattern as above.

**Review by crolarlibertyva — CHANGES_REQUESTED** [review]

**Review by bdgrider — APPROVED** [review]
Reviewed changes, verified all unit and regression tests are passing.

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js (hunk 1 above)
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js:40]: IDE recommends the following change.

> Comparison error.response != undefined may cause unexpected type coercion 
> Comparison error.response.status == 400 may cause unexpected type coercion

`error.response !== undefined && error.response.status === 400`

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js (hunk 2 above)
Comment (beaugrantham-va) [inline oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js:54]: IDE recommends the following change.

> Comparison error.response != undefined may cause unexpected type coercion 
> Comparison error.response.status == 401 may cause unexpected type coercion

`error.response !== undefined && error.response.status === 401`

File: oauth-proxy/tests/TokenHandlerTests/getDocumentByAccessTokenStrategy.test.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      dynamoClient,
+      config,
+      hashingFunction
+    ).getDocument("access_token");
+
+    expect(document.access_token.S).toBe("access_token");
+    expect(document.launch.S).toBe("launch");
+  });
+
+  it("Dynamo Error", async () => {
```
Comment (beaugrantham-va) [inline oauth-proxy/tests/TokenHandlerTests/getDocumentByAccessTokenStrategy.test.js:38]: This test name should be more descriptive (it's testing a specific error scenario where it's looking up an access_token that does not exist).

File: oauth-proxy/tests/TokenHandlerTests/getDocumentByLaunchStrategy.test.js [hunk 4]
```diff
@@ -1,27 +1,29 @@
...
 require("jest");
 
 describe("getDocumentByLaunchStrategy tests", () => {
   it("undefined launch", async () => {
-    const req = {
-      body: {},
-    };
+    const req = new MockExpressRequest({
+      body: {
+        launch: undefined,
```
Comment (beaugrantham-va) [inline oauth-proxy/tests/TokenHandlerTests/getDocumentByLaunchStrategy.test.js:12]: I think this should be a separate test. The previous scenario `body: {}` is valid and should still be covered.

File: oauth-proxy/tests/TokenHandlerTests/authorizationCodeStrategy.test.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
+const { buildFakeLogger, buildFakeOktaClient } = require("../testUtils");
+const Collection = require("@okta/okta-sdk-nodejs/src/collection");
+const ModelFactory = require("@okta/okta-sdk-nodejs/src/model-factory");
+const User = require("@okta/okta-sdk-nodejs/src/models/User");
+
+const getAuthorizationServerInfoMock = jest.fn();
+
+const userCollection = new Collection("", "", new ModelFactory(User));
```
Comment (beaugrantham-va) [inline oauth-proxy/tests/TokenHandlerTests/authorizationCodeStrategy.test.js:8]: IDE reports the following warning for the first param here.

> Argument type string is not assignable to parameter type ApiClient

**Review by beaugrantham-va — CHANGES_REQUESTED** [review]

File: oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js (hunk 1 above)
Comment (lukemarkwordtlibertyits) [reply oauth-proxy/oauthHandlers/tokenHandlerStrategyClasses/tokenStrategies/clientCredentialsStrategy.js]: Check for a "truthy" value is the desired result

**Review by crolarlibertyva — APPROVED** [review]

**Review by dbrown-va — APPROVED** [review]
Confirmed changes implemented from Beau's earlier review.
